spopt/tests/test_locate.py

```diff
@@ -0,0 +1,531 @@
+
+    def test_optimality_pmedian_from_cost_matrix(self):
+        pmedian = PMedian.from_cost_matrix(
+            self.cost_matrix, self.ai, p_facilities=self.p_facility
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        self.assertEqual(pmedian.problem.status, pulp.LpStatusOptimal)
+
+    def test_infeasibility_pmedian_from_cost_matrix(self):
+        pmedian = PMedian.from_cost_matrix(self.cost_matrix, self.ai, p_facilities=0)
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        self.assertEqual(pmedian.problem.status, pulp.LpStatusInfeasible)
+
+    def test_pmedian_facility_client_array_from_cost_matrix(self):
+        with open(self.dirpath + "pmedian_objective.pkl", "rb") as f:
+            pmedian_objective = pickle.load(f)
+
+        pmedian = PMedian.from_cost_matrix(
+            self.cost_matrix, self.ai, p_facilities=self.p_facility
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        pmedian.facility_client_array()
+
+        self.assertCountEqual(pmedian.fac2cli, pmedian_objective)
+
+    def test_mixin_mean_distance(self):
+        mean_distance_expected = 2982.1268579890657
+        pmedian = PMedian.from_cost_matrix(
+            self.cost_matrix, self.ai, p_facilities=self.p_facility
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        pmedian.get_mean_distance(self.ai)
+
+        self.assertEqual(pmedian.mean_dist, mean_distance_expected)
+
+    def test_optimality_pmedian_from_geodataframe(self):
+        pmedian = PMedian.from_geodataframe(
+            self.demand_points_gdf,
+            self.facility_points_gdf,
+            "geometry",
+            "geometry",
+            "POP2000",
+            p_facilities=self.p_facility,
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        self.assertEqual(pmedian.problem.status, pulp.LpStatusOptimal)
+
+    def test_infeasibility_pmedian_from_geodataframe(self):
+        pmedian = PMedian.from_geodataframe(
+            self.demand_points_gdf,
+            self.facility_points_gdf,
+            "geometry",
+            "geometry",
+            "POP2000",
+            p_facilities=0,
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        self.assertEqual(pmedian.problem.status, pulp.LpStatusInfeasible)
+
+    def test_pmedian_facility_client_array_from_geodataframe(self):
+        with open(self.dirpath + "pmedian_geodataframe_objective.pkl", "rb") as f:
+            pmedian_objective = pickle.load(f)
+
+        pmedian = PMedian.from_geodataframe(
+            self.demand_points_gdf,
+            self.facility_points_gdf,
+            "geometry",
+            "geometry",
+            "POP2000",
+            p_facilities=self.p_facility,
+        )
+        pmedian = pmedian.solve(pulp.PULP_CBC_CMD())
+        pmedian.facility_client_array()
+
+        self.assertCountEqual(pmedian.fac2cli, pmedian_objective)
```

Comment:
[`numpy.testing.assert_array_equal()`](https://github.com/pysal/spopt/blob/42520cc53d10c14a61ac8ba79bdd45e080281775/spopt/tests/test_maxp.py#L88) is very useful for testing equality of iterables.